storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -4010,6 +3994,33 @@ dberr_t fil_space_t::garbage_collect(bool shutdown)
     return err;
   }
 
+  /* Reset the undo log segment slots in the rollback segment header
+  which exists in system tablespace. Undo cached segment will be
+  treated as unused file segment. These segment will be freed as a
+  part of inode_info::free_segs  */
+  mtr.start();
+  mtr.x_lock_space(fil_system.sys_space);
+  for (trx_rseg_t &rseg : trx_sys.rseg_array)
+  {
+    if (rseg.space == fil_system.sys_space &&
+        UT_LIST_GET_LEN(rseg.undo_cached))
+    {
+      buf_block_t *block=
+        buf_page_get_gen(page_id_t{0, rseg.page_no}, 0,
+                         RW_X_LATCH, nullptr, BUF_GET, &mtr, &err);
+      if (!block)
+      {
+        mtr.commit();
+        return err;
+      }
+
+      mtr.memset(block, TRX_RSEG_UNDO_SLOTS + TRX_RSEG,
+                 TRX_RSEG_N_SLOTS * TRX_RSEG_SLOT_SIZE, 0xff);
```

Comment:
I checked that without this, the test would fail like this:
```diff
CURRENT_TEST: innodb.undo_leak
--- /home/marko/11.2/mysql-test/suite/innodb/r/undo_leak.result	2024-11-15 06:37:39.933929352 -0800
+++ /home/marko/11.2/mysql-test/suite/innodb/r/undo_leak.reject	2024-11-15 06:50:26.151920244 -0800
@@ -43,4 +43,4 @@
 # restart
 SELECT NAME, FILE_SIZE FROM information_schema.innodb_sys_tablespaces WHERE SPACE = 0;
 NAME	FILE_SIZE
-innodb_system	12582912
+innodb_system	18874368
```